os/mm/mm_heap/mm_heapinfo_utils.c

```diff
@@ -0,0 +1,201 @@
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <sys/types.h>
+#include <tinyara/sched.h>
+#include <tinyara/mm/mm.h>
+
+
+/****************************************************************************
+ * Name: heapinfo_add_size
+ *
+ * Description:
+ * Add the allocated size in tcb
+ ****************************************************************************/
+void heapinfo_add_size(struct mm_heap_s *heap, pid_t pid, mmsize_t size)
+{
+	pid_t hash_pid;
+
+	hash_pid = PIDHASH(pid);
+	if (heap->alloc_list[hash_pid].pid == HEAPINFO_INIT_INFO || heap->alloc_list[hash_pid].pid == pid) {
+			heap->alloc_list[hash_pid].pid = pid;
+			heap->alloc_list[hash_pid].curr_alloc_size += size;
+			if (heap->alloc_list[hash_pid].curr_alloc_size > heap->alloc_list[hash_pid].peak_alloc_size) {
+				heap->alloc_list[hash_pid].peak_alloc_size = heap->alloc_list[hash_pid].curr_alloc_size;
+			}
+			heap->alloc_list[hash_pid].num_alloc_free++;
+	}
+}
+
+/****************************************************************************
+ * Name: heapinfo_subtract_size
+ *
+ * Description:
+ * Subtract the allocated size in tcb
+ ****************************************************************************/
+void heapinfo_subtract_size(struct mm_heap_s *heap, pid_t pid, mmsize_t size)
+{
+	pid_t hash_pid;
+
+	hash_pid = PIDHASH(pid);
+	if (heap->alloc_list[hash_pid].pid == pid) {
+			heap->alloc_list[hash_pid].curr_alloc_size -= size;
+			heap->alloc_list[hash_pid].num_alloc_free--;
+	}
+}
+
+/****************************************************************************
+ * Name: heapinfo_update_total_size
+ *
+ * Description:
+ * Calculate the total allocated size and update the peak allocated size for heap
+ ****************************************************************************/
+void heapinfo_update_total_size(struct mm_heap_s *heap, mmsize_t size, pid_t pid)
+{
+	heap->total_alloc_size += size;
+	if (heap->total_alloc_size > heap->peak_alloc_size) {
+		heap->peak_alloc_size = heap->total_alloc_size;
+	}
+#if CONFIG_KMM_NHEAPS > 1
+	total_info.cur_alloc_size += size;
+	if (total_info.cur_alloc_size > total_info.peak_alloc_size) {
+		total_info.peak_alloc_size = total_info.cur_alloc_size;
+	}
+#endif
+#ifdef CONFIG_HEAPINFO_USER_GROUP
+	heapinfo_update_group(size, pid);
+#endif
+}
+/****************************************************************************
+ * Name: heapinfo_update_node
+ *
+ * Description:
+ * Adds pid and malloc caller return address to mem chunk
+ ****************************************************************************/
+void heapinfo_update_node(FAR struct mm_allocnode_s *node, mmaddress_t caller_retaddr)
+{
+	node->alloc_call_addr = caller_retaddr;
```

Comment:
node is always valid? same comment for all below functions.

Reply:
Yes, it should be. But I will add debugassert for debugging. Thanks.

Reply:
Ok but I mentioned heapinfo_exclude_stacksize & heapinfo_peak_init too.